overlord/snapstate/backend/copydata.go

```diff
@@ -333,3 +333,64 @@ func (b Backend) RemoveExposedSnapHome(snapName string) error {
 
 	return firstErr
 }
+
+// XDGDirsUndoInfo contains information about what XDG dirs were moved or created
+// in InitXDGDirs. This allows the caller to undo the operation.
+type XDGDirsUndoInfo struct {
+	Moved   []string
+	Created []string
+}
+
+// InitXDGDirs renames .local/share, .config and .cache directories to their
+// post core22 migration locations. Directories that don't exist are created.
+// The returned XDGDirsUndoInfo struct can be used to undo this operation.
+func (b Backend) InitXDGDirs(info *snap.Info) (undoInfo *XDGDirsUndoInfo, err error) {
+	opts := &dirs.SnapDirOptions{HiddenSnapDataDir: true, MigratedToExposedHome: true}
+	users, err := allUsers(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	undoInfo = &XDGDirsUndoInfo{}
+
+	srcXdgDirs := []string{".config", ".cache", ".local/share"}
+	dstXdgDirs := []string{"xdg-config", "xdg-cache", "xdg-data"}
+
+	for _, usr := range users {
+		for i, srcDir := range srcXdgDirs {
+			revDir := info.UserDataDir(usr.HomeDir, opts)
+			src := filepath.Join(revDir, srcDir)
+			dst := filepath.Join(revDir, dstXdgDirs[i])
+
+			if exists, isDir, err := osutil.DirExists(src); err != nil {
+				return undoInfo, err
+			} else if exists && isDir {
+				if err := os.Rename(src, dst); err != nil {
+					return undoInfo, err
+				}
+
+				undoInfo.Moved = append(undoInfo.Moved, dst)
+
+				// since we copied the contents of the revisioned dir to ~/Snap previously,
+				// remove the XDG directories in there
+				exposedXdgDir := filepath.Join(info.UserExposedHomeDir(usr.HomeDir), srcDir)
+				if err := os.RemoveAll(exposedXdgDir); err != nil {
+					return undoInfo, err
+				}
+			} else {
+				uid, gid, err := osutil.UidGid(usr)
+				if err != nil {
+					return undoInfo, err
+				}
+
+				if err := mkdirAllChown(dst, 0700, uid, gid); err != nil {
+					return undoInfo, err
```

Comment:
I'm wondering here if a malicious snap could cause the process to fail, if it creates a file (and not a directory) `$SNAP_USER_DATA/<snap>/xdg-config` before this migration happens.

If it is expected that the migration would fail (there's not much we can do, admittedly), then maybe it would be good to use `fmt.Errorf` and combine the `err` with some text that clearly says that we couldn't recreate `dst` as a directory (I looked at the implementation of `mkdirAllChown()` and I'm afraid that errors reported by it won't be very clean on their own.

Reply:
Actually, I think this applies to `os.Rename()` above too.

Reply:
Yes, this one of the corner cases that I wrote in the spec file for migration. There are other places where an existing file can break the process as well. That's also why the new xdg dirs have the "xdg-" prefix, so clashes are less likely.

Reply:
Added a check and a test for the case where the destination already exists. Thanks